File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n=nums.length;
        int k=n-2;
        int l=n-1;

        while(k>=0 && nums[k]>=nums[k+1])
        {
            k--;
        }
        if(k==-1)
        {
           reverse(nums,0,n-1);
           return;
        }

        while(l>k && nums[l]<=nums[k])
        {
            l--;
        }

        swap(nums,k,l);
        reverse(nums,k+1,n-1);
        
    }

    public void reverse(int nums[],int k,int l)
    {
       while(k<l)
       {
           swap(nums,k,l);
           k++;
           l--;

       }
    }

     public void swap(int nums[],int k,int l)
    {
       int temp=nums[k];
       nums[k]=nums[l];
       nums[l]=temp;
    }
}

File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k = n - 2
        l = n - 1

        
        while k >= 0 and nums[k] >= nums[k + 1]:
            k -= 1
        
        if k == -1:
            self.reverse(nums, 0, n - 1)
            return

       
        while l > k and nums[l] <= nums[k]:
            l -= 1

        self.swap(nums, k, l)
        self.reverse(nums, k + 1, n - 1)

    def swap(self, nums, i, j):
        nums[i], nums[j] = nums[j], nums[i]

    def reverse(self, nums, start, end):
        while start < end:
            self.swap(nums, start, end)
            start += 1
            end -= 1
